components/service/firefox-accounts/src/main/java/mozilla/components/service/fxa/manager/SyncEnginesStorage.kt

```diff
@@ -6,12 +6,17 @@ package mozilla.components.service.fxa.manager
 
 import android.content.Context
 import mozilla.components.service.fxa.SyncEngine
+import mozilla.components.service.fxa.store.SyncAction
+import mozilla.components.service.fxa.store.SyncStore
 import mozilla.components.service.fxa.sync.toSyncEngine
 
 /**
  * Storage layer for the enabled/disabled state of [SyncEngine].
  */
-class SyncEnginesStorage(private val context: Context) {
+class SyncEnginesStorage(
+    private val context: Context,
+    private val syncStore: SyncStore
```

Comment:
> This should be a required change. Not all consumers would want to use the store.

This comment is with the assumption that "consumer" means consumer of code within Fenix. If it is more generally "consumer" of A-C, then I agree. 

The underlying issue is we need a way to guarantee that changes to the `SyncEnginesStorage` propagates to the `SyncStore`. If we don't require this parameter, then we can't guarantee that. We can update every call in A-C and Fenix that we care about, but nothing would stop client code from constructing a storage without it and making updates. I'm not sure I see a good way around this, open to feedback.

One alternative would be to switch to an observable persistence layer, like DataStore or Room. That feels like a much larger lift though.

I am okay if, for now, we want to do it as you've suggested but I worry it opens the door to future bugs.

Reply:
> The underlying issue is we need a way to guarantee that changes to the `SyncEnginesStorage` propagates to the `SyncStore`. If we don't require this parameter, then we can't guarantee that. We can update every call in A-C and Fenix that we care about, but nothing would stop client code from constructing a storage without it and making updates. I'm not sure I see a good way around this, open to feedback.

I think this that is an acceptable API to have. Forcing all consumers (non-Fenix) to setup a SyncStore that is not also a required way to use the AccountManager looks wrong. Looking at examples in our codebase, in the past, when we wanted to migrate from `SessionManager` to `BrowserStore`, we added [a similar nullable param](https://github.com/mozilla-mobile/android-components/blob/v69.0.0/components/browser/session/src/main/java/mozilla/components/browser/session/SessionManager.kt#L31) to have one system "sync" to other, but did not enforce _all consumers_ to follow our requirements. This also makes it easier for us to land this in Fenix this way, so that we can land these API changes, and migrate to Fenix when ready without it causing a breaking API and requiring an immediate patch.

If that logic is fine, then we want to focus this down to how do we make sure Fenix usages have the SyncStore, we could add an annotation on the field to draw attention that if there is no instance set, then the user has to specifically suppress it.

[Here is an example](https://github.com/mozilla-mobile/android-components/blob/54c57f88d77d6af07edb6a778275b5b62700d418/components/browser/state/src/main/java/mozilla/components/browser/state/action/BrowserAction.kt#L1406) where I used the `@DelicateApi` annotation so that callers of this action read the docs when using it and have to take explicit care to suppress the annotation. We can create a simple annotation that does the same thing (DelicateApi doesn't make sense to reuse here). I also don't have strong requirements for going about it in this way specifically (this will cause failures by warningsAsErrors), we can try others or don't do anything for it as well.